test/functional/mining_basic.py

```diff
@@ -131,5 +138,67 @@ def run_test(self):
         bad_block.hashPrevBlock = 123
         assert_template(node, bad_block, 'inconclusive-not-best-prevblk')
 
+        self.log.info('submitheader tests')
+        assert_raises_rpc_error(-22, 'Block header decode failed', lambda: node.submitheader(hexdata='xx' * 80))
+        assert_raises_rpc_error(-22, 'Block header decode failed', lambda: node.submitheader(hexdata='ff' * 78))
+        assert_raises_rpc_error(-25, 'Must submit previous header', lambda: node.submitheader(hexdata='ff' * 80))
+
+        block.solve()
+
+        def chain_tip(b_hash, *, status='headers-only', branchlen=1):
+            return {'hash': b_hash, 'height': 202, 'branchlen': branchlen, 'status': status}
+
+        assert chain_tip(block.hash) not in node.getchaintips()
+        node.submitheader(hexdata=b2x(block.serialize()))
+        assert chain_tip(block.hash) in node.getchaintips()
+        node.submitheader(hexdata=b2x(CBlockHeader(block).serialize()))  # Noop
+        assert chain_tip(block.hash) in node.getchaintips()
+
+        bad_block_root = copy.deepcopy(block)
+        bad_block_root.hashMerkleRoot += 2
+        bad_block_root.solve()
+        assert chain_tip(bad_block_root.hash) not in node.getchaintips()
+        node.submitheader(hexdata=b2x(CBlockHeader(bad_block_root).serialize()))
+        assert chain_tip(bad_block_root.hash) in node.getchaintips()
+        # Should still reject invalid blocks, even if we have the header:
+        assert_equal(node.submitblock(hexdata=b2x(bad_block_root.serialize())), 'invalid')
+        assert chain_tip(bad_block_root.hash) in node.getchaintips()
+        # We know the header for this invalid block, so should just return early without error:
+        node.submitheader(hexdata=b2x(CBlockHeader(bad_block_root).serialize()))
+        assert chain_tip(bad_block_root.hash) in node.getchaintips()
+
+        bad_block_lock = copy.deepcopy(block)
+        bad_block_lock.vtx[0].nLockTime = 2**32 - 1
+        bad_block_lock.vtx[0].rehash()
+        bad_block_lock.hashMerkleRoot = bad_block_lock.calc_merkle_root()
+        bad_block_lock.solve()
+        assert_equal(node.submitblock(hexdata=b2x(bad_block_lock.serialize())), 'invalid')
```

Comment:
Does master have a bug? This should be returning "bad-txns-nonfinal", not "invalid"...

Reply:
See the TODO https://github.com/bitcoin/bitcoin/blob/8803c9132a78d8182bd828a29f7051fc7688f934/src/rpc/mining.cpp#L756